src/index/hnsw/faiss_hnsw.cc

```diff
@@ -0,0 +1,1398 @@
+                k >= (index->ntotal - filtered_out_num) * kHnswSearchBFTopkThreshold) {
+                return true;
+            }
+        }
+
+        // the default value
+        return false;
+    }
+
+    Status
+    AddInternal(const DataSetPtr dataset, const Config&) override {
+        if (index == nullptr) {
+            LOG_KNOWHERE_ERROR_ << "Can not add data to an empty index.";
+            return Status::empty_index;
+        }
+
+        auto rows = dataset->GetRows();
+        try {
+            LOG_KNOWHERE_INFO_ << "Adding " << rows << " to HNSW Index";
+
+            auto status = add_to_index(index.get(), dataset, data_format);
+            if (status != Status::success) {
+                return status;
+            }
+
+        } catch (const std::exception& e) {
+            LOG_KNOWHERE_WARNING_ << "faiss inner error: " << e.what();
+            return Status::faiss_inner_error;
+        }
+
+        return Status::success;
+    }
+};
+
+//
+class BaseFaissRegularIndexHNSWFlatNode : public BaseFaissRegularIndexHNSWNode {
+ public:
+    BaseFaissRegularIndexHNSWFlatNode(const int32_t& version, const Object& object, DataFormatEnum data_format)
+        : BaseFaissRegularIndexHNSWNode(version, object, data_format) {
+    }
+
+    bool
+    HasRawData(const std::string& metric_type) const override {
+        if (index == nullptr) {
+            return false;
+        }
+
+        // yes, a flat index has it
+        return true;
+    }
+
+    std::unique_ptr<BaseConfig>
+    CreateConfig() const override {
+        return std::make_unique<FaissHnswFlatConfig>();
+    }
+
+    std::string
+    Type() const override {
+        return knowhere::IndexEnum::INDEX_FAISS_HNSW_FLAT;
+    }
+
+ protected:
+    Status
+    TrainInternal(const DataSetPtr dataset, const Config& cfg) override {
+        // number of rows
+        auto rows = dataset->GetRows();
+        // dimensionality of the data
+        auto dim = dataset->GetDim();
+        // data
+        auto data = dataset->GetTensor();
+
+        // config
+        auto hnsw_cfg = static_cast<const FaissHnswFlatConfig&>(cfg);
+
+        auto metric = Str2FaissMetricType(hnsw_cfg.metric_type.value());
+        if (!metric.has_value()) {
+            LOG_KNOWHERE_ERROR_ << "Invalid metric type: " << hnsw_cfg.metric_type.value();
+            return Status::invalid_metric_type;
+        }
+
+        // create an index
+        const bool is_cosine = IsMetricType(hnsw_cfg.metric_type.value(), metric::COSINE);
+
+        std::unique_ptr<faiss::IndexHNSW> hnsw_index;
+        if (is_cosine) {
+            if (data_format == DataFormatEnum::fp32) {
+                hnsw_index = std::make_unique<faiss::IndexHNSWFlatCosine>(dim, hnsw_cfg.M.value());
+            } else if (data_format == DataFormatEnum::fp16) {
+                hnsw_index = std::make_unique<faiss::IndexHNSWSQCosine>(dim, faiss::ScalarQuantizer::QT_fp16,
```

Comment:
Yes and no.
Adding data into index consists of two steps:
1. training an index
2. adding the data.

Training HNSW, HNSWFlat, HNSWSQ(fp16) and HNSWSQ(bf16) does literally nothing. So, line [763](https://github.com/zilliztech/knowhere/blob/094fa745ef20472ff6f4c7a5fc5f79a81be0c44a/src/index/hnsw/faiss_hnsw.cc#L763)
`hnsw_index->train(rows, (const float*)data);` does literally nothing and is just a formality, as mentioned in the comment on lines 758-762.
 
Adding the data is performed via `add_to_index()` function on [line 346](https://github.com/zilliztech/knowhere/blob/094fa745ef20472ff6f4c7a5fc5f79a81be0c44a/src/index/hnsw/faiss_hnsw.cc#L346). If the input data format is not `fp32`, then it will split the input data into virtual chunks of `65536` lines ([line 356](https://github.com/zilliztech/knowhere/blob/094fa745ef20472ff6f4c7a5fc5f79a81be0c44a/src/index/hnsw/faiss_hnsw.cc#L356)). Every chunk will be converted into fp32 and added into an index. So, the overhead is `65536 * dim` bytes. If you believe that this is a large value, then please let me know what a better value should be. `16384`? `8192`? even `4096` maybe? `4096` for a `1536 dim` OpenAI dataset is just extra `24 MB`

Reply:
Also, this memory overconsumption is a problem for HNSWPQ and HNSWPRQ at this moment, because Faiss processed everything in floats. But there is a relatively simple way to fix it in the future because Faiss has all the needed things in place, thus trading training speed for the memory, if needed (we can make it a config parameter).

Reply:
changed to 4096

Reply:
IC, so after calling `add_to_index()`, FAISS HNSWSQ(BF16) will only store BF16 data right?

Also, can you help elaborate a little bit why HNSWPQ/PRQ need more memory. Actually building memory is not that critical, but search is